syntax = "proto2";
package hw.trezor.messages.aptos;

// Sugar for easier handling in Java
option java_package = "com.satoshilabs.trezor.lib.protobuf";
option java_outer_classname = "TrezorMessageAptos";

/**
 * Request: Address at the specified index
 * @start
 * @next AptosAddress
 */
message AptosGetAddress {
    repeated uint32 address_n = 1;              // BIP-32 path to derive the key from master node
    optional bool show_display = 2;             // optionally show on display before sending the result
}

/**
 * Response: Address for the given index
 * @end
 */
message AptosAddress {
    optional string address = 1;                // Aptos address as hex-encoded string
}

/**
 * Request: ask device to sign Aptos transaction
 * @start
 * @next AptosSignedTx
 */
message AptosSignTx {
    repeated uint32 address_n = 1;         // BIP-32 path to derive the key from master node
    required bytes raw_tx = 2;              // serialized raw transaction
}

/**
 * Response: signature for transaction
 * @end
 */
message AptosSignedTx {
    required bytes public_key = 1;  // public key for the private key used to sign tx
    required bytes signature = 2;           // the signature of the raw transaction
}

/**
 * Request: ask device to sign Aptos message
 * @start
 * @next AptosMessageSignature
 */
message AptosSignMessage {
    repeated uint32 address_n = 1;
    required AptosMessagePayload payload = 2;

    message AptosMessagePayload {
        optional string address = 2;
        optional string chain_id = 3;
        optional string application = 4;
        required string nonce = 5;
        required string message = 6;
    }
}

/**
 * Response: signature for message
 * @end
 */
message AptosMessageSignature {
    required bytes signature = 1;
    required string address = 2;
}
